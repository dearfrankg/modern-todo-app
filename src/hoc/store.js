import React, { Component } from "react";
import { connect } from "react-redux";

export default function(ComposedComponent, reducers) {
  class decoreWithStore extends Component {
    componentWillMount() {
      // eslint-disable-next-line
      // console.log('Will Mount HOC', reducers);
    }

    render() {
      return <ComposedComponent {...this.props} />;
    }
  }

  function mapStateToProps(state) {
    // eslint-disable-next-line
    if (reducers.length === 0) console.error("There' no reducer here, fix it!", reducers);

    const reducerMap = {};

    reducers.forEach(reducer => {
      reducerMap[reducer] = state[reducer];
    });

    return reducerMap;
  }

  return connect(mapStateToProps)(decoreWithStore);
}
