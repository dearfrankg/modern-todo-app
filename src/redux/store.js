import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import rootReducer from 'redux/reducers';
import { loadState, saveState } from 'redux/localStorage';

const persistedState = loadState();

const store = createStore(
  combineReducers({
    todos: rootReducer,
    router: routerReducer,
  }),
  persistedState,
);

store.subscribe(() => {
  saveState({
    todos: store.getState().todos,
  });
});

export default store;
