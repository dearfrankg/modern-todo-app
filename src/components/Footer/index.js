import React from 'react';
import {
  arrayOf, any, func, string,
} from 'prop-types';
import { Link } from 'react-router-dom';

import { getItemsLeftLabel, getListStatus } from 'helpers';
import * as filters from 'redux/types';

const propTypes = {
  todos: arrayOf(any).isRequired,
  clearCompleted: func.isRequired,
  activeFilter: string.isRequired,
};

const Footer = ({ todos, clearCompleted, activeFilter }) => {
  const hasTodos = todos.length > 0;
  const hasCompletedItems = getListStatus(todos).completedItems > 0;

  const footerComponent = (
    <div>
      <span className="todo-count">
        {getItemsLeftLabel(todos)}
      </span>

      <ul className="filters">
        <li>
          <Link className={activeFilter !== filters.SHOW_ALL ? '' : 'selected'} to="#/">
            All
          </Link>
        </li>
        <li>
          <Link className={activeFilter !== filters.SHOW_ACTIVE ? '' : 'selected'} to="#/active">
            Active
          </Link>
        </li>
        <li>
          <Link
            className={activeFilter !== filters.SHOW_COMPLETED ? '' : 'selected'}
            to="#/completed"
          >
            Completed
          </Link>
        </li>
      </ul>

      {hasCompletedItems && (
        <button className="clear-completed" onClick={clearCompleted}>
          Clear completed
        </button>
      )}
    </div>
  );

  return (
    <footer className="footer">
      {hasTodos && footerComponent}
    </footer>
  );
};

Footer.propTypes = propTypes;

export default Footer;
