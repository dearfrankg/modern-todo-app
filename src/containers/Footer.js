import React from 'react';
import { objectOf, any } from 'prop-types';
import { connect } from 'react-redux';

import { Footer } from 'components';
import { clearCompleted } from 'redux/actions';
import { getFilterFromPath } from 'helpers';
import withStore from 'hoc';

const propTypes = {
  router: objectOf(any).isRequired,
};

export const FooterContainer = props => (
  <Footer {...props} activeFilter={getFilterFromPath(props.router.location.hash || '#/')} />
);

export const mapDispatchToProps = dispatch => ({
  clearCompleted: () => dispatch(clearCompleted()),
});

FooterContainer.propTypes = propTypes;

export default connect(
  null,
  mapDispatchToProps,
)(withStore(FooterContainer, ['todos', 'router']));
