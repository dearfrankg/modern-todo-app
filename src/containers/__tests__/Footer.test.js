import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';

import store from 'redux/store';
import Footer, { FooterContainer, mapDispatchToProps } from '../Footer';

describe('Footer Container:', () => {
  it('should render correctly with default props', () => {
    expect(
      shallow(
        <Provider store={store}>
          <Footer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with default props', () => {
    const wrapper = mount(
      <FooterContainer
        router={{ location: { hash: '#/' } }}
        clearCompleted={() => {}}
        todos={[]}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with default props [2]', () => {
    const wrapper = mount(
      <FooterContainer router={{ location: { hash: '' } }} clearCompleted={() => {}} todos={[]} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('should call actions', () => {
    const dispatchSpy = jest.fn();
    const { clearCompleted } = mapDispatchToProps(dispatchSpy);

    it('should call clearCompleted', () => {
      clearCompleted();
      const spyLastCall = dispatchSpy.mock.calls[0][0];
      expect(spyLastCall.type).toEqual(dispatchSpy.mock.calls[0][0].type);
    });
  });
});
