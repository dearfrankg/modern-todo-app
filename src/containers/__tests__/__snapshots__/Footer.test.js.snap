// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Container: should render correctly with default props [2] 1`] = `
<FooterContainer
  clearCompleted={[Function]}
  router={
    Object {
      "location": Object {
        "hash": "",
      },
    }
  }
  todos={Array []}
>
  <Footer
    activeFilter="show_all"
    clearCompleted={[Function]}
    router={
      Object {
        "location": Object {
          "hash": "",
        },
      }
    }
    todos={Array []}
  >
    <footer
      className="footer"
    />
  </Footer>
</FooterContainer>
`;

exports[`Footer Container: should render correctly with default props 1`] = `<Connect(Connect(decoreWithStore)) />`;

exports[`Footer Container: should render correctly with default props 2`] = `
<FooterContainer
  clearCompleted={[Function]}
  router={
    Object {
      "location": Object {
        "hash": "#/",
      },
    }
  }
  todos={Array []}
>
  <Footer
    activeFilter="show_all"
    clearCompleted={[Function]}
    router={
      Object {
        "location": Object {
          "hash": "#/",
        },
      }
    }
    todos={Array []}
  >
    <footer
      className="footer"
    />
  </Footer>
</FooterContainer>
`;
